### PR TITLE
Resolves #787: Support for custom Executor per FDBRecordContext

### LINKED ISSUE #787: Expose functionality similar to ContextRestoringExecutor
When an `FDBRecordContext` is opened, it takes the `Executor` that is returned from the `FDBDatabase` and wraps it in a `ContextRestoringExecutor`.  The `ContextRestoringExecutor` is used to support the way in which we log--the current thread has a `ThreadLocal` map of name/value pairs that will automatically be appended to every log message and it is the job of the `ContextRestoringExecutor` to ensure that this map is visible as a `ThreadLocal` in the executor thread that picks up and runs a piece of Record Layer work.

This behavior would be nice to expose as a publicly available piece of functionality by applications that also use thread locals to maintain application (for the record, this is generally a tremendously bad design decision but, alas, a strategy that some applications have already invested in).  

This could take the form of adding an alternative form of:

    FDBDatabaseFactory.setExecutor()

as

    FDBDatabaseFactory.setExecutor(Supplier<Executor> executorGenerator);

Then the provided generator can do whatever it wants to set up the executor prior to returning it.

### PR DESCRIPTION
**Pull request by scgray**

### CODE REVIEW

File: fdb-extensions/src/main/java/com/apple/foundationdb/async/TaskNotifyingExecutor.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    public TaskNotifyingExecutor(@Nonnull Executor executor) {
+        this.delegate = executor;
+    }
+
+    @Override
+    public void execute(final Runnable task) {
+        if (task instanceof Notifier) {
+            delegate.execute(task);
+        } else {
+            delegate.execute(new Notifier( task));
```
Comment (MMcM) [inline fdb-extensions/src/main/java/com/apple/foundationdb/async/TaskNotifyingExecutor.java]:
```suggestion
            delegate.execute(new Notifier(task));
```

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBDatabaseFactory.java [hunk 2]
```diff
@@ -67,8 +67,13 @@
 
     @Nullable
     private Executor networkExecutor = null;
+
+    @Nonnull
     private Executor executor = ForkJoinPool.commonPool();
 
+    @Nonnull
+    private Function<Executor, Executor> contextExecutor = exec -> exec;
```
Comment (MMcM) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBDatabaseFactory.java]:
```suggestion
    private Function<Executor, Executor> contextExecutor = Function.identity();
```

Comment (MMcM) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBDatabaseFactory.java]: (I believe IntelliJ would offer the same suggestion.)

File: fdb-extensions/src/main/java/com/apple/foundationdb/async/TaskNotifyingExecutor.java (hunk 1 above)
Comment (scgray) [reply fdb-extensions/src/main/java/com/apple/foundationdb/async/TaskNotifyingExecutor.java]: Done

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBDatabaseFactory.java (hunk 2 above)
Comment (scgray) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBDatabaseFactory.java]: It does, but I was on the fence about whether or not I preferred the more explicit syntax (seeing how `Function.identity()` simply returns `x -> x`).   Anyway, updated.

**Pull request merged**